### PR TITLE
Feat: Study API

### LINKED ISSUE #15: feat: studies
- [ ] [Export one study chapter](https://lichess.org/api#tag/Studies) - studies.export({ studyId, chapterId })
- [ ] [Export all chapters](https://lichess.org/api#tag/Studies/operation/studyAllChaptersPgn) - studies.export({ studyId })
- [ ] [Export all studies of a user](https://lichess.org/api#tag/Studies/operation/studyExportAllPgn) - studies.export({ username })
- [ ] [Study metadata](https://lichess.org/api#tag/Studies/operation/studyAllChaptersHead) - studies.metadata()
- [ ] [List studies of a user](https://lichess.org/api#tag/Studies/operation/studyListMetadata) - user.studies() vs studies.user()

### PR DESCRIPTION
**Pull request by nicholasdejong**
- tweak: Fetcher now supports all REST methods
- Slight tweaks. May remove in the future.
- Feat: Study API

### what & why

Fixes #15. The lichess Study API is now supported, but testing and documentation still need to be added.
Note: `user.studies` is responsible for streaming a list of the user's studies. Since streaming is currently not supported, this API is provisional until streaming is implemented.

In order to accomodate more REST functionality, I slightly tweaked the `fetcher` API, replacing `post?: boolean` with `method?: string`. This allows using the fetcher to GET, POST, UPDATE, DELETE, HEAD, etc. as opposed to just GET and POST. This tweak was necessary to accomodate the new Study API, and will also be required for future APIs. Let me know how you feel about the tweak.

### how

New Study APIs:
- [x]  `study.chapter({ studyId, chapterId })`: Export a chapter with id `chapterId` from study with id `studyId`.
- [x] `study.chapters({ studyId })`: Export all chapters from study with id `studyId`.
- [x] `study.studies({ username })`: Export all studies from user with username `username`. 
- [x] `study.meta({ studyId })`: Retrieve metadata from study with id `studyId`.
- [x] `study.import({ studyId, name, pgn })`: Create a chapter with PGN `pgn` and name `name` in study with id `studyId`.
- [ ] `user.studies({ username })`: Stream all studies of a user with username `username` (provisional).

### CODE REVIEW

**Comment by nicholasdejong** [comment]
TODO: impl Study API documentation. Will get to it soon, but want to fix streams first.

**Comment by devjiwonchoi** [comment]
Thank you for the PR! This project is using `pnpm`, please run `pnpm i` instead of `npm`!

File: package.json [hunk 1]
```diff
@@ -6,8 +6,8 @@
   "types": "dist/index.d.ts",
   "license": "MIT",
   "scripts": {
-    "build": "bunchee",
-    "test": "NODE_ENV=test dotenv -e .env jest"
+    "build": "npx bunchee",
```
Comment (devjiwonchoi) [inline package.json:9]: No need `npx`, bunchee is in the devDeps.

```suggestion
    "build": "bunchee",
```

File: .gitignore [hunk 2]
```diff
@@ -2,4 +2,5 @@ dist
 node_modules
 .env
 .DS_Store
-.next
\ No newline at end of file
+.next
+package-lock.json
```
Comment (devjiwonchoi) [inline .gitignore:6]: Run `pnpm i`, not `npm i`
```suggestion
```

File: package.json [hunk 3]
```diff
@@ -47,6 +47,6 @@
   ],
   "engines": {
     "node": ">=18.17.0",
-    "pnpm": "8.9.0"
+    "pnpm": ">=8.9.0"
```
Comment (devjiwonchoi) [inline package.json:50]: We could upgrade pnpm version on a separate PR, the version is fixed for easier reproduction.

File: src/utils.ts [hunk 4]
```diff
@@ -3,20 +3,20 @@ import { LICHESS_API_URL } from './constants'
...
+  method?: string
   body?: URLSearchParams
 }) {
   return (
     await fetch(`${LICHESS_API_URL}${endpoint}`, {
       headers: {
         Authorization: `Bearer ${token}`,
       },
-      method: post ? 'POST' : 'GET',
+      method,
```
Comment (devjiwonchoi) [inline src/utils.ts:19]: We can set the default method to `GET`, so we only need to pass other methods if necessary

**Review by devjiwonchoi — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (nicholasdejong) [reply package.json:9]: That is well then. I needed npx on my machine but that's okay.

File: .gitignore (hunk 2 above)
Comment (nicholasdejong) [reply .gitignore:6]: Apologies, still new to pnpm.

File: package.json [hunk 5]
```diff
@@ -7,7 +7,7 @@
   "license": "MIT",
   "scripts": {
     "build": "bunchee",
-    "test": "NODE_ENV=test dotenv -e .env jest"
+    "test": "NODE_ENV=test dotenv -e .env jest --silent=false"
```
Comment (devjiwonchoi) [inline package.json:10]: Why do we need this?

```suggestion
    "test": "NODE_ENV=test dotenv -e .env jest"
```

**Comment by devjiwonchoi** [comment]
Thank you! Please add test cases for studies on the `test` dir!

**Comment by nicholasdejong** [comment]
I get the 'Missing Scope' Error frequently when  trying  to run  the tests myself.  Are the tests automatically tested for each pr?

**Comment by devjiwonchoi** [comment]
Take a look at this for testing: https://github.com/devjiwonchoi/equine/blob/main/contributing/testing.md
You need a lichess dev API token

**Review by devjiwonchoi — APPROVED** [review]

**Comment by devjiwonchoi** [comment]
Yeah will take a look, but we do need tests though. The tests passes on local.

**Comment by nicholasdejong** [comment]
Working on tests right now.

**Review by devjiwonchoi — DISMISSED** [review]

**Comment by nicholasdejong** [comment]
The tests don't really work on my side. The fetch API tends to fail on me. In order to test the creation of a chapter we need to create a study, but I can't find the API for that on the Study reference.

**Comment by devjiwonchoi** [comment]
Thank you!

**Pull request merged**